kani-compiler/src/codegen_aeneas_llbc/mir_to_ullbc/mod.rs

```diff
@@ -0,0 +1,797 @@
+/// The results of the translation are stored in the `translated` field.
+pub struct Context<'a, 'tcx> {
+    tcx: TyCtxt<'tcx>,
+    instance: Instance,
+    translated: &'a mut TranslatedCrate,
+    errors: &'a mut ErrorCtx<'tcx>,
+}
+
+impl<'a, 'tcx> Context<'a, 'tcx> {
+    /// Create a new context for translating the function `instance`, populating
+    /// the results of the translation in `translated`
+    pub fn new(
+        tcx: TyCtxt<'tcx>,
+        instance: Instance,
+        translated: &'a mut TranslatedCrate,
+        errors: &'a mut ErrorCtx<'tcx>,
+    ) -> Self {
+        Self { tcx, instance, translated, errors }
+    }
+
+    fn tcx(&self) -> TyCtxt<'tcx> {
+        self.tcx
+    }
+
+    fn span_err<S: Into<MultiSpan>>(&mut self, span: S, msg: &str) {
+        self.errors.span_err(span, msg);
+    }
+
+    fn continue_on_failure(&self) -> bool {
+        self.errors.continue_on_failure()
+    }
+
+    /// Perform the translation
+    pub fn translate(&mut self) -> Result<(), ()> {
+        // Charon's `id_map` is in terms of internal `DefId`
+        let def_id = rustc_internal::internal(self.tcx(), self.instance.def.def_id());
+
+        // TODO: might want to populate `errors.dep_sources` to help with
+        // debugging
+
+        let fid = self.register_fun_decl_id(def_id);
+
+        let item_meta = match self.translate_item_meta_from_rid(self.instance) {
+            Ok(item_meta) => item_meta,
+            Err(_) => {
+                return Err(());
+            }
+        };
+
+        let signature = self.translate_function_signature();
+        let body = match self.translate_function_body() {
+            Ok(body) => body,
+            Err(_) => {
+                return Err(());
+            }
+        };
+
+        let fun_decl = FunDecl {
+            def_id: fid,
+            rust_id: def_id,
+            item_meta,
+            signature,
+            kind: ItemKind::Regular,
+            body: Ok(body),
+        };
+
+        self.translated.fun_decls.set_slot(fid, fun_decl);
+
+        Ok(())
+    }
+
+    /// Get or create a `FunDeclId` for the given function
+    fn register_fun_decl_id(&mut self, def_id: InternalDefId) -> FunDeclId {
+        let tid = match self.translated.id_map.get(&def_id) {
+            Some(tid) => *tid,
+            None => {
+                let tid = AnyTransId::Fun(self.translated.fun_decls.reserve_slot());
+                self.translated.id_map.insert(def_id, tid);
+                self.translated.reverse_id_map.insert(tid, def_id);
+                self.translated.all_ids.insert(tid);
+                tid
+            }
+        };
+        *tid.as_fun()
+    }
+
+    /// Compute the meta information for a Rust item identified by its id.
+    fn translate_item_meta_from_rid(&mut self, instance: Instance) -> Result<ItemMeta, Error> {
+        let span = self.translate_instance_span(instance);
+        let name = self.def_id_to_name(instance.def.def_id())?;
```

Comment:
I'm confused. Why can't we use `instance.mangled_name()`?

Reply:
My understanding is that this tries to preserve the original names as much as possible so that it's easy for the user to relate the generated code to the original Rust.

Reply:
I think it would be great to get more clarity on why we need this extra complexity. I'm also concerned that this could impact the analysis soundness (in case of 2+ names being resolved to the same name). It's also not clear to me how this is handling generic parameters.

Reply:
Another potential issue is with different crate versions.. I wonder if they will have the same name for multiple items.

Reply:
These names are used in two ways: as @zhassan-aws said we use them when displaying info to the user so they can recognize which item we're talking about, and we also sometimes pattern-match on them to recognize built-ins. As of latest charon we use lang/diagnostic items wherever possible, but we still rely on names to recognize e.g. `core::panicking::panic_explicit`. That's not crucial.

Generic parameters are only relevant for inherent impl blocks. We include them in the name:
```rust
impl<T: Clone> Foo<T> {
    fn method(&self) {}
}
```
This method has name `crate::module::{impl<T: Clone> Foo<T>}::method`.

Well spotted on crate versions, I opened an [issue](https://github.com/AeneasVerif/charon/issues/412) for that.